pm: make PM_DEVICE_RUNTIME independent of PM_DEVICE

gmarull opened the pull request:
There's no need to enable CONFIG_PM_DEVICE to have device runtime PM, and this has actually led to a lot of confusion by Zephyr developers. Make the necessary changes so that CONFIG_PM_DEVICE_RUNTIME can be enabled standalone.  This allows to enable the CONFIG_PM + CONFIG_PM_DEVICE_RUNTIME on a system without worrying about the problems associated with CONFIG_PM_DEVICE.

Note that this PR still keeps a coupling between device.h/device_runtime.h, could be eliminated entirely, this PR is just one step towards isolating CONFIG_PM_DEVICE so that developers finally understand what it means.

NOTE: I had to drop 1560bfe, tree needs a big cleanup before this can go in (device PM is so broken).

gmarull committed:
pm: make device runtime PM idependent from PM_DEVICE

There's no need to enable CONFIG_PM_DEVICE to have device runtime PM,
and this has actually led to a lot of confusion by Zephyr developers.
Make the necessary changes so that CONFIG_PM_DEVICE_RUNTIME can be
enabled standalone.  This allows to enable the CONFIG_PM +
CONFIG_PM_DEVICE_RUNTIME on a system without worrying about the problems
associated with CONFIG_PM_DEVICE.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
pm: device: clarify which APIs are exclusive to PM_DEVICE

Some APIs are not meant to be used with runtime device PM, make sure
guards are placed correctly.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: pinctrl: sleep state is also allowed for device runtime PM

This change will allow operating without enabling CONFIG_PM_DEVICE.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
pm: drop PM_DEVICE_RUNTIME_EXCLUSIVE

It no longer makes sense, now that runtime PM can be enabled without
enabling CONFIG_PM_DEVICE.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

JordanYates commented:
Broadly speaking this effort seems like a good idea, I have used device runtime PM for years without ever using the system power  management functionality.

gmarull committed:
pm: device_runtime: introduce new resource definition helpers

Add specific helpers for device runtime PM. This will allow decoupling
PM device from PM device runtime. With a single macro for both _models_,
one must always compile with PM_DEVICE, otherwise, drivers not
implementing runtime PM would fail to compile if only enabling
PM_DEVICE_RUNTIME.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

knthm commented:
Frankly, this seems like a step backward from final the goal of deprecating `PM_DEVICE` entirely, at least in comparison to #60939.

gmarull committed:
pm: drivers using runtime PM must use runtime helpers

Instead of using helpers from PM_DEVICE.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
> Broadly speaking this effort seems like a good idea, I have used device runtime PM for years without ever using the system power management functionality.

You've been lucky because Nordic SoCs don't deal with CONFIG_PM (now it's not even selectable, thanks to `sys_poweroff()`).

ceolin reviewed (changes requested):
It is a good idea allows one just using DEVICE_RUNTIME without PM_DEVICE, but I didn't get why it is need a new set of macros to initialize pm in a device, that is code duplication. Drop runtime exclusive is part of https://github.com/zephyrproject-rtos/zephyr/pull/60939

gmarull commented:
> Add specific helpers for device runtime PM. This will allow decoupling
> PM device from PM device runtime. With a single macro for both _models_,
> one must always compile with PM_DEVICE, otherwise, drivers not
> implementing runtime PM would fail to compile if only enabling
> PM_DEVICE_RUNTIME.

It's explained in the commit message:

> Add specific helpers for device runtime PM. This will allow decoupling
PM device from PM device runtime. With a single macro for both _models_,
one must always compile with PM_DEVICE, otherwise, drivers not
implementing runtime PM would fail to compile if only enabling
PM_DEVICE_RUNTIME.

ceolin commented:
> > Add specific helpers for device runtime PM. This will allow decoupling
> > PM device from PM device runtime. With a single macro for both _models_,
> > one must always compile with PM_DEVICE, otherwise, drivers not
> > implementing runtime PM would fail to compile if only enabling
> > PM_DEVICE_RUNTIME.
> 
> It's explained in the commit message:
> 
> > Add specific helpers for device runtime PM. This will allow decoupling
> > PM device from PM device runtime. With a single macro for both _models_,
> > one must always compile with PM_DEVICE, otherwise, drivers not
> > implementing runtime PM would fail to compile if only enabling
> > PM_DEVICE_RUNTIME.

it is not clear for me, sorry. If you have selected only `PM_DEVICE_RUNTIME` and not implemented device runtime, isn't it supposed to fail ?

>> otherwise, drivers not
> > implementing runtime PM would fail to compile if only enabling
> > PM_DEVICE_RUNTIME.

gmarull commented:
> > > Add specific helpers for device runtime PM. This will allow decoupling
> > > PM device from PM device runtime. With a single macro for both _models_,
> > > one must always compile with PM_DEVICE, otherwise, drivers not
> > > implementing runtime PM would fail to compile if only enabling
> > > PM_DEVICE_RUNTIME.
> > 
> > 
> > It's explained in the commit message:
> > > Add specific helpers for device runtime PM. This will allow decoupling
> > > PM device from PM device runtime. With a single macro for both _models_,
> > > one must always compile with PM_DEVICE, otherwise, drivers not
> > > implementing runtime PM would fail to compile if only enabling
> > > PM_DEVICE_RUNTIME.
> > > > one must always compile with PM_DEVICE, otherwise, drivers not
> > > implementing runtime PM would fail to compile if only enabling
> > > PM_DEVICE_RUNTIME.
> 
> it is not clear for me, sorry. If you have selected only `PM_DEVICE_RUNTIME` and not implemented device runtime, isn't it supposed to fail ?
> 
> > > otherwise, drivers not
> > > implementing runtime PM would fail to compile if only enabling
> > > PM_DEVICE_RUNTIME.

> it is not clear for me, sorry. If you have selected only `PM_DEVICE_RUNTIME` and not implemented device runtime, isn't it supposed to fail ?

You can still select CONFIG_PM + CONFIG_PM_DEVICE + CONFIG_PM_DEVICE_RUNTIME. Since this is possible, separate macros need to exist. Otherwise, if CONFIG_PM_DEVICE=n drivers implementing CONFIG_PM_DEVICE (so they guard the action callback with CONFIG_PM_DEVICE) will fail to compile because the PM macros will expand and expect a valid callback.

ceolin commented:
> > > > Add specific helpers for device runtime PM. This will allow decoupling
> > > > PM device from PM device runtime. With a single macro for both _models_,
> > > > one must always compile with PM_DEVICE, otherwise, drivers not
> > > > implementing runtime PM would fail to compile if only enabling
> > > > PM_DEVICE_RUNTIME.
> > > 
> > > 
> > > It's explained in the commit message:
> > > > Add specific helpers for device runtime PM. This will allow decoupling
> > > > PM device from PM device runtime. With a single macro for both _models_,
> > > > one must always compile with PM_DEVICE, otherwise, drivers not
> > > > implementing runtime PM would fail to compile if only enabling
> > > > PM_DEVICE_RUNTIME.
> > > > > one must always compile with PM_DEVICE, otherwise, drivers not
> > > > > implementing runtime PM would fail to compile if only enabling
> > > > > PM_DEVICE_RUNTIME.
> > 
> > 
> > it is not clear for me, sorry. If you have selected only `PM_DEVICE_RUNTIME` and not implemented device runtime, isn't it supposed to fail ?
> > > > otherwise, drivers not
> > > > implementing runtime PM would fail to compile if only enabling
> > > > PM_DEVICE_RUNTIME.
> 
> > it is not clear for me, sorry. If you have selected only `PM_DEVICE_RUNTIME` and not implemented device runtime, isn't it supposed to fail ?
> 
> You can still select CONFIG_PM + CONFIG_PM_DEVICE + CONFIG_PM_DEVICE_RUNTIME. Since this is possible, separate macros need to exist. Otherwise, if CONFIG_PM_DEVICE=n drivers implementing CONFIG_PM_DEVICE (so they guard the action callback with CONFIG_PM_DEVICE) will fail to compile because the PM macros will expand and expect a valid callback.

Gotcha, the problem is the combination of all these symbols. Still, having a new set of macros seems to be overkill, will play around it a little bit more.

carlescufi reviewed (changes requested):
Please update the migration guide with the required changes in applications

gmarull commented on doc/services/pm/device.rst:
```diff
@@ -67,12 +67,6 @@ option for device power management.
 
-.. note::
-
-    This type of device power management is disabled when
```
`CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE` is a kind of `CONFIG_PM_DEVICE=n`, so if you're using it, you don't want `CONFIG_PM_DEVICE`, just that you can't because device runtime depends on it. When using both, if it's the case, PM_DEVICE will still skip devices using runtime PM.

ceolin commented:
@gmarull can you explain exactly what it this pr goal ? Being able to enable only device runtime ? I think we can to it in a simple way.  PM_DEVICE_RUNTIME_EXCLUSIVE is a confuse name and when it is set we make runtime pm and device pm mutually exclusive. We can de-couple them making a new symbol, e.g `SYSTEM_PM_DEVICE`, that when is not set we skip device pm from the sleep process. We make `PM_DEVICE` a hidden symbol that is selected if `PM_DEVICE_RUNTIME || SYSTEM_PM_DEVICE` and that contains the shared infra (macros and struct fields).

The changes in this pr makes devices implemented using device runtime not compatible with system device pm, breaking the current functionality. Looking the drivers changed in this pr is not clear if any of these devices have any problem with system device pm model ...  if the majority of devices can use both models, we should let users decide which one they want to use, especially because there is also a cost associated with device runtime pm. Devices that are compatible with only one model can use the appropriate guard when defining pm data and callback if needed, or query the system at runtime ....

gmarull commented:
closing, no longer an active contributor to the PM subsys.
